app/assets/javascripts/discourse/app/app.js

```diff
@@ -67,15 +70,82 @@ const Discourse = Application.extend({
     }
 
     Object.keys(requirejs._eak_seen).forEach((key) => {
-      if (/\/pre\-initializers\//.test(key)) {
-        const initializer = this._prepareInitializer(key);
-        if (initializer) {
-          this.initializer(initializer);
-        }
-      } else if (/\/(api\-)?initializers\//.test(key)) {
-        const initializer = this._prepareInitializer(key);
-        if (initializer) {
-          this.instanceInitializer(initializer);
+      if (key.startsWith("discourse/")) {
```

Comment:
I think it would be good to split this up into two PRs. It seems like adding this `key.startsWith()` check to our existing initializer logic should be sufficient to solve the compatibility problem with the `ember-functions-as-helper-polyfill` addon?

Introducing support for `/instance-initializers/` in Discourse core/themes/plugins does seem reasonable, but I would be more comfortable doing that in a separate PR, with a plan in place to deprecate 'discourse-style' initializers.

How does that sound?

Reply:
(There is the one in its `addon/` folder that should never be run in the first place that would be fixed by the discourse prefix check, which should always have been there for correctness, but it re-exports the same initializer in the `app/` folder which merges it into the discourse modules namespace, which is how v1 addons installs stuff into apps)

Reply:
#22095